docs/config.md
## Configuration

You can configure everything via plain environment variables, `.env` or `config.toml`. Precedence of configuration: env > config.

| Description                                  | Env                  | Default               | Note                                        |
|:---------------------------------------------|:---------------------|:----------------------|:--------------------------------------------|
| Server port                                  | `PORT`               | 8080                  | Do not change when using docker compose     |
| Prefix for admin API<sup>1</sup>             | `ADMIN_API`          | -                     | Optional. If not provided, API is disabled  |
| If should allow new registers                | `ALLOW_NEW_REGISTER` | `true`                |                                             |
| Limit for JSON payload for requests          | `LIMITS_JSON`        | 4MiB<sup>2</sup>      | Original server has no limit                |
| Secret text for encoding/decoding JWT tokens | `JWT_SECRET`         | -                     | Required                                    |
| JWT issuer                                   | `JWT_ISSUER`         | `http://0.0.0.0:8080` | Do not add `/` at the end                   |
| Path to database file                        | `DATABASE_URL`       | data.db               | For SQLite                                  |
| Name of database in MySQL                    | `DATABASE_NAME`      | kotatsu_db            | For MySQL                                   |
| Host of MySQL database                       | `DATABASE_HOST`      | localhost             | For MySQL                                   |
| Port of MySQL database                       | `DATABASE_PORT`      | 3306                  | For MySQL                                   |
| User for connecting to MySQL database        | `DATABASE_USER`      | -                     | For MySQL. Required                         |
| Password for user in MySQL database          | `DATABASE_PASSWORD`  | -                     | For MySQL. Required                         |
| Log level                                    | `RUST_LOG`           | `error`<sup>3</sup>   |                                             |

1. Enables some additiional features, like statistics. For `/admin` URL will look like `http://IP/admin/stats`
1. Examples: 256 kB, 0.500 mib, 1MB, 1GiB
1. Possible values: `off`, `error`, `warn`, `info`, `debug`, `trace`. In debug build default `info`

### Example `.env`

```bash
# mysql
DATABASE_NAME=kotatsu_db
DATABASE_HOST=localhost
DATABASE_PORT=3306
DATABASE_USER=YOUR_USER
DATABASE_PASSWORD=YOUR_PASSWORD

# sqlite
DATABASE_URL=data.db

PORT=8080
JWT_SECRET=SECRET
JWT_ISSUER=http://0.0.0.0:8080
ALLOW_NEW_REGISTER=true
ADMIN_API=/admin
LIMITS_JSON=4MiB
```

### Example `config.toml`

Currently name of config file can't be configured.

```toml
[server]
port = 8080
admin_api = "/admin"
allow_new_register = true

[server.limits]
json = "4MiB"

[db]
# sqlite
url = "data.db"

# mysql
name = "kotatsu_db"
host = "localhost"
port = 3306
user = "YOUR_USER"
password = "YOUR_PASSWORD"

[jwt]
secret = ""
issuer = "http://0.0.0.0:8080"
```


CHANGELOG.md
# Changelog

## Unreleased

- RUST_LOG can now be set in `.env`

## v0.3.0-beta.1 (2025-09-28)

### Added

- Server can now be built as a static binary
- JWT issuer can be configured with `JWT_ISSUER` env now

### Updated

- Update to latest changes (up to Kotatsu 9.2)

### Removed

- JWT audience cannot be configured now

## v0.2 (2024-08-16)

### Added

- Configure port (env `PORT`)
- Configure limit for JSON requests payloads (env `LIMITS_JSON`)

### Changed

- Config structure has been changed

## v0.1 (2024-06-01)

Initial release. Includes all features from original server. Also includes:

- Use SQLite for storage
- Use and migrate MySQL database from original server
- Uses Blake3 for hashing passwords
- Can be configured via config file
- `get /manga`: max `limit` is 1000


src/config.rs
use std::fmt::Display;

use rocket::data::ByteUnit;

// When add new values here also update docs/docker-compose.yaml and
// docs/config.md
#[derive(Debug, Clone, confique::Config)]
pub struct Conf {
    #[config(nested)]
    pub server: ConfServer,
    #[config(nested)]
    pub db: ConfDB,
    #[config(nested)]
    pub jwt: ConfJWT,
}

#[derive(Debug, Clone, confique::Config)]
pub struct ConfServer {
    #[config(env = "PORT", default = 8080)]
    pub port: u16,
    #[config(env = "ADMIN_API")]
    pub admin_api: Option<String>,
    #[config(env = "ALLOW_NEW_REGISTER", default = true)]
    pub allow_new_register: bool,
    #[config(nested)]
    pub limits: ConfServerLimits,
}

#[derive(Debug, Clone, confique::Config)]
pub struct ConfServerLimits {
    #[config(env = "LIMITS_JSON", default = "4MiB")]
    pub json: ByteUnit,
}

#[derive(Debug, Clone, confique::Config)]
pub struct ConfJWT {
    #[config(env = "JWT_SECRET")]
    pub secret: String,
    #[config(env = "JWT_ISSUER", default = "http://0.0.0.0:8080")]
    pub issuer: String,
}

#[cfg(feature = "sqlite")]
#[derive(Debug, Clone, confique::Config)]
pub struct ConfDB {
    #[config(env = "DATABASE_URL", default = "data.db")]
    pub url: String,
}

#[cfg(feature = "mysql")]
#[derive(Debug, Clone, confique::Config)]
pub struct ConfDB {
    #[config(env = "DATABASE_NAME", default = "kotatsu_db")]
    pub name: String,
    #[config(env = "DATABASE_HOST", default = "localhost")]
    pub host: String,
    #[config(env = "DATABASE_PORT", default = 3306)]
    pub port: u16,
    #[config(env = "DATABASE_USER")]
    pub user: String,
    #[config(env = "DATABASE_PASSWORD")]
    pub password: String,
}

impl ConfDB {
    #[cfg(feature = "sqlite")]
    pub fn url(&self) -> String {
        self.url.clone()
    }

    #[cfg(feature = "mysql")]
    pub fn url(&self) -> String {
        format!(
            "mysql://{user}:{password}@{host}:{port}/{name}",
            user = self.user,
            password = self.password,
            host = self.host,
            port = self.port,
            name = self.name,
        )
    }
}

impl Display for Conf {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.pad("Config\n  ")?;

        f.pad("server.port: ")?;
        self.server.port.fmt(f)?;
        f.pad("\n  server.admin_api: ")?;
        if let Some(admin_api) = &self.server.admin_api {
            f.pad(admin_api)?;
        } else {
            f.pad("[empty]")?;
        }
        f.pad("\n  server.allow_new_register: ")?;
        self.server.allow_new_register.fmt(f)?;
        f.pad("\n  server.limits.json: ")?;
        self.server.limits.json.fmt(f)?;

        f.pad("\n  db.url: ")?;
        f.pad(&self.db.url())?;

        f.pad("\n  jwt.secret: ")?;
        if self.jwt.secret.is_empty() {
            f.pad("[empty]")?;
        } else {
            f.pad("[redacted]")?;
        }
        f.pad("\n  jwt.issuer: ")?;
        f.pad(&self.jwt.issuer)?;

        Ok(())
    }
}
